File: leetcode/ways-to-make-a-fair-array/myHan.java
class Solution {
    public int waysToMakeFair(int[] nums) {
        //idx => 0, 1, [2, 3, ...] l + 2, l + 3
        int[] acc = new int[nums.length+4];
        int[] racc = new int[nums.length+4];
        
        for(int i = 2; i <= nums.length; i++) {
            acc[i] = acc[i-2] + nums[i-2];
        }
        for(int i = nums.length + 1; 2 <= i; i--) {
            racc[i] = racc[i+2] + nums[i-2];
        }
        int ret = 0;
        for(int i = 2; i < nums.length + 2; i++) {
            if(acc[i-1] + racc[i+2] == acc[i-2] + racc[i+1]) ret++;
        }
        return ret;
    }
}

File: leetcode/15-maximum-gap/myHan.scala
import scala.language.postfixOps

object Solution {
    def maximumGap(nums: Array[Int]): Int = {
        var order = 0 until nums.size toArray
        
        for(lb <- 0 until 30 by 6) {
            val pat = (1 << (lb + 6)) - 1
            val count = Array.fill[Int](1<<6)(0)
            for(i <- order) {
                count((pat & nums(i)) >> lb) += 1
            }
            
            for(i <- 1 until count.size) {
                count(i) += count(i-1)
            }
            val nOrder = new Array[Int](order.size)
            for(i <- order.size-1 to 0 by -1) {
                count((nums(order(i)) & pat) >> lb) -= 1
                nOrder(count((nums(order(i)) & pat) >> lb)) = order(i)
            }
            order = nOrder
        }
        
        var ret = 0
        for(i <- 0 until order.size-1) {
            ret = ret max (nums(order(i+1)) - nums(order(i)))
        }
        ret
    }
}